libr/bin/format/mach0/coresymbolication.c

```diff
@@ -0,0 +1,368 @@
+			goto beach;
+		}
+		size_t i;
+		ut8 *cursor = b + R_CS_EL_OFF_SEGS;
+		for (i = 0; i < hdr->n_segments && cursor < end; i++) {
+			RCoreSymCacheElementSegment *seg = &result->segments[i];
+			seg->paddr = seg->vaddr = r_read_le64 (cursor);
+			cursor += 8;
+			if (cursor >= end) {
+				break;
+			}
+			seg->size = seg->vsize = r_read_le64 (cursor);
+			cursor += 8;
+			if (cursor >= end) {
+				break;
+			}
+			seg->name = str_dup_safe_fixed (b, cursor, 16, end);
+			cursor += 16;
+			if (!seg->name) {
+				continue;
+			}
+
+			if (!strcmp (seg->name, "__PAGEZERO")) {
+				page_zero_size = seg->size;
+				page_zero_idx = i;
+				seg->paddr = seg->vaddr = 0;
+				seg->size = 0;
+			}
+		}
+		for (i = 0; i < hdr->n_segments && page_zero_size > 0; i++) {
+			if (i == page_zero_idx) {
+				continue;
+			}
+			RCoreSymCacheElementSegment *seg = &result->segments[i];
+			if (seg->vaddr < page_zero_size) {
+				seg->vaddr += page_zero_size;
+			}
+		}
+	}
+	const ut64 start_of_sections = (ut64)hdr->n_segments * R_CS_EL_SIZE_SEG + R_CS_EL_OFF_SEGS;
+	if (hdr->n_sections > 0) {
+		result->sections = R_NEWS0 (RCoreSymCacheElementSection, hdr->n_sections);
+		if (!result->sections) {
+			goto beach;
+		}
+		size_t i;
+		ut8 *cursor = b + start_of_sections;
+		for (i = 0; i < hdr->n_sections && cursor < end; i++) {
+			ut8 *sect_start = cursor;
+			RCoreSymCacheElementSection *sect = &result->sections[i];
+			sect->vaddr = sect->paddr = r_read_ble (cursor, false, bits);
+			if (sect->vaddr < page_zero_size) {
+				sect->vaddr += page_zero_size;
+			}
+			cursor += word_size;
+			if (cursor >= end) {
+				break;
+			}
+			sect->size = r_read_ble (cursor, false, bits);
+			cursor += word_size;
+			if (cursor >= end) {
+				break;
+			}
+			ut64 sect_name_off = r_read_ble (cursor, false, bits);
+			cursor += word_size;
+			if (bits == 32) {
+				cursor += word_size;
+			}
+			sect->name = str_dup_safe (b, sect_start + (size_t)sect_name_off, end);
+		}
+	}
+	const ut64 sect_size =  (bits == 32) ? R_CS_EL_SIZE_SECT_32 : R_CS_EL_SIZE_SECT_64;
+	const ut64 start_of_symbols = start_of_sections + (ut64)hdr->n_sections * sect_size;
+	if (hdr->n_symbols) {
+		result->symbols = R_NEWS0 (RCoreSymCacheElementSymbol, hdr->n_symbols);
+		if (!result->symbols) {
+			goto beach;
+		}
+		size_t i;
+		ut8 *cursor = b + start_of_symbols;
+		for (i = 0; i < hdr->n_symbols && cursor + R_CS_EL_SIZE_SYM <= end; i++) {
+			RCoreSymCacheElementSymbol *sym = &result->symbols[i];
+			sym->paddr = r_read_le32 (cursor);
+			sym->size = r_read_le32 (cursor + 0x4);
+			sym->unk1 = r_read_le32 (cursor + 0x8);
+			size_t name_off = r_read_le32 (cursor + 0xc);
+			size_t mangled_name_off = r_read_le32 (cursor + 0x10);
+			sym->unk2 = (st32)r_read_le32 (cursor + 0x14);
+			sym->name = str_dup_safe (b, cursor + name_off, end);
+			if (!sym->name) {
+				break;
+			}
+			sym->mangled_name = str_dup_safe (b, cursor + mangled_name_off, end);
+			if (!sym->mangled_name) {
+				break;
```

Comment:
it is a problem if one of those (name, mangled_name) is not a valid address? can we just skip it instead of break?

Reply:
we can probably miss some hits on corrupted or tweaked files